### PR TITLE
MAP-112_Enable employer to respond to OLTD request on COMPLETED record

### PR DESCRIPTION
**Pull request by ratheeshri1984**

### CODE REVIEW

**Review by cofaulco — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/ApprenticeController.cs [hunk 1]
```diff
@@ -970,6 +979,34 @@ public IActionResult ConfirmHasNotStopChanges(ConfirmHasNotStopViewModel viewMod
...
+                    await _commitmentsApiClient.ResolveOverlappingTrainingDateRequest(new ResolveApprenticeshipOverlappingTrainingDateRequest
+                    {
+                        ApprenticeshipId = viewModel.ApprenticeshipId,
+                        ResolutionType = OverlappingTrainingDateRequestResolutionType.ApprenticeshipEndDateUpdate
+                    }, CancellationToken.None);
+
+                    TempData.AddFlashMessage(ApprenticeEndDateConfirmed, ITempDataDictionaryExtensions.FlashMessageLevel.Success);
+
+                    return RedirectToAction(nameof(ApprenticeshipDetails), new ApprenticeshipDetailsRequest { AccountHashedId = viewModel.AccountHashedId, ApprenticeshipHashedId = viewModel.ApprenticeshipHashedId });
+
```
[the remark this answers is not in the document]
Comment (ratheeshri1984) [reply src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/ApprenticeController.cs:999]: Done

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/ApprenticeController.cs [hunk 2]
```diff
@@ -970,6 +979,34 @@ public IActionResult ConfirmHasNotStopChanges(ConfirmHasNotStopViewModel viewMod
             return View(viewModel);
         }
 
+        [HttpPost]
+        [Route("{apprenticeshipHashedId}/details/confirmHasValidEndDate")]
+        public async Task<IActionResult> ConfirmHasValidEndDateChanges(ConfirmHasValidEndDateViewModel viewModel)
+        {
+            if (viewModel.EndDateConfirmed.HasValue)
+            {
```
[the remark this answers is not in the document]
Comment (ratheeshri1984) [reply src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/ApprenticeController.cs:987]: Done

File: src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Controllers/ApprenticeControllerTests/WhenRequestingConfirmApprenticeshipHasValidEndDatePage.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
+﻿using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.Logging;
+using Moq;
+using NUnit.Framework;
+using SFA.DAS.CommitmentsV2.Api.Client;
+using SFA.DAS.CommitmentsV2.Shared.Interfaces;
+using SFA.DAS.EmployerCommitmentsV2.Web.Controllers;
+using SFA.DAS.EmployerCommitmentsV2.Web.Models.Apprentice;
+using SFA.DAS.Testing.AutoFixture;
+using System.Threading.Tasks;
```
[the remark this answers is not in the document]
Comment (ratheeshri1984) [reply src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Controllers/ApprenticeControllerTests/WhenRequestingConfirmApprenticeshipHasValidEndDatePage.cs:10]: Done

File: src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Controllers/ApprenticeControllerTests/WhenPostingConfirmHasValidEndDateChanges.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+using Microsoft.AspNetCore.Mvc.ViewFeatures;
+using Microsoft.Extensions.Logging;
+using Moq;
+using NUnit.Framework;
+using SFA.DAS.CommitmentsV2.Api.Client;
+using SFA.DAS.CommitmentsV2.Shared.Interfaces;
+using SFA.DAS.EmployerCommitmentsV2.Web.Controllers;
+using SFA.DAS.EmployerCommitmentsV2.Web.Models.Apprentice;
+using SFA.DAS.Testing.AutoFixture;
+using System.Threading.Tasks;
```
[the remark this answers is not in the document]
Comment (ratheeshri1984) [reply src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Controllers/ApprenticeControllerTests/WhenPostingConfirmHasValidEndDateChanges.cs:13]: Done

File: src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Controllers/ApprenticeControllerTests/WhenRequestingConfirmApprenticeshipHasValidEndDatePage.cs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+                _mockCookieStorageService.Object,
+                _mockCommitmentsApiClient.Object,
+                Mock.Of<ILogger<ApprenticeController>>());
+        }
+
+        [Test, MoqAutoData]
+        public async Task WhenRequesting_ConfirmApprenticeshipHasValidEndDate_ThenConfirmHasValidEndDateRequestViewModelIsPassedToTheView(ConfirmHasValidEndDateViewModel expectedViewModel)
+        {
+            _mockModelMapper
+                .Setup(m => m.Map<ConfirmHasValidEndDateViewModel>(It.IsAny<ConfirmHasValidEndDateRequest>()))
```
[the remark this answers is not in the document]
Comment (ratheeshri1984) [reply src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Controllers/ApprenticeControllerTests/WhenRequestingConfirmApprenticeshipHasValidEndDatePage.cs]: Done

**Review by cofaulco — APPROVED** [review]

**Pull request merged**